recipe/meta.yaml

```diff
@@ -25,7 +25,7 @@ requirements:
     - ipywidgets >=7.0.0,<8
     - jupyter-cache >=0.4.1,<0.5
     - jupyter-sphinx ==0.3.1
-    - myst-parser ~=0.13.3
+    - myst-parser >=0.13.3,<0.14.0
```

Comment:
aren't these two things equivalent? why change it? we use the `~=` syntax throughout all of the other `executablebooks` repositories so this would be a break from convention

Reply:
despite whatever EB does, i see a fair number of these in the wild that are just plain _wrong_. meanwhile, across the more than 10k conda-forge repositories, where we ship stuff from _all over_ that _works together_, the `>` syntax is widely used because it's:
- understood by humans from a wider swathe of educational backgrounds (e.g. second or third year maths)
  - `~` and `^` have _wildly_ different meanings in other settings
- specifies an actual, numeric upper and lower bounds helps people when they get unsolveable `conda` errors or `pip check` failures, rather than presenting them with a puzzle

Reply:
> despite whatever EB does

I appreciate your help with these distributions. But it is an EBP practice to use `~`, and a perfectly legitimate part of the requirement specification.
The requirements here should directly replicate those found in the main repository (setup.cfg), and I'm going to have to put my foot down here, and say that is the only practice that I am accepting

Reply:
I think there's a reasonable case to be made in either direction on the specific syntax. Since these conda-forge recipes are likely going to have participation from folks in the EBP world, I think there's a strong value in just having consistency for how things are done. If the syntax should change, I think that decision should be made at the community-level rather than deciding in ad-hoc fashion in repositories here and there

Reply:
The issue is more that putting upper bounds *at all* causes problems across conda packages because it makes it harder to solve for dependencies in any given environment. EBP may have very strict protocols about following semantic versioning but most projects don't, so conda forge has to work with the lowest common denominator in general, i.e. don't assume version numbers are semantic versioned, only that they are in sequence. The general problem with upper bounds is that you can't predict the future and know a priori if that version really is an upper bound. The best method is to go back and add upper bounds to old conda packages after a new version is released and the upper bound is definitively known. The `~` syntax seems to be mostly tied to the idea of semantic versioning and thus isn't such a nice fit for conda dependency trees that span all packages in an environment.

This article does a good job at explaining the issues with semantic versioning: https://hynek.me/articles/semver-will-not-save-you/ if you are curious.

The best thing to do is to leave the upper bounds off on new releases. But, if you really want to predict upper bounds on new releases, then making them explicit is better than not, as per @bollyyvl's suggestions. And this is because conda packages have to be as widely compatible with other packages in an environment as possible, not just EBP's ecosystem of packages.

Many, many conda recipes deviate the build specifications from the build specs in the project's own testing environment to maximize compatibility. Same goes with debian, fedora, homebrew, etc.

Reply:
> This article does a good job at explaining the issues with semantic versioning:

yes I have read this and also all the discussion around it: https://news.ycombinator.com/item?id=26314620

Reply:
> The issue is more that putting upper bounds at all causes problems across conda packages because it makes it harder to solve for dependencies in any given environment.

yep thats great; it **should** be harder to solve; we are trying to specify an environment where the package will actually work; it is almost definite that it will not work with new major versions of packages, given that they will be breaking in nature.
What is the point of being able to solve for an environment that does not actually work?

I would strongly suggest that you are trying to "fix" an issue that has never existed; I can't think of any bugs that have been opened in EBP about a faulty conda distribution,
but I could point out mny, many times these would have broken if we had allowed for a newer version of a package to be installed:

- Every package in the jupyter-book stack would have broken when I released markdown-it-py v1
- myst-nb would break on every major release of myst-parser
- Jupyter-book would break when we merge this PR

...